### PR TITLE
improve resultCallback handling

### PR DESCRIPTION
**Pull request by fengxx**
ResultCallbackTemplate.close doesn't work under some circumstances: for example
some command may take a long time to get a http response, 
1) launch a container with `sleep 1h`
2) tail the logs,  `GET /v1.24/containers/ID/logs?tail=-1&since=0&stdout=true&stderr=true&follow=true` is sent and it will take 1 hour to get a response.

If client tried to abort the tail by calling close like

```
completed=logCallback.awaitCompletion(30, TimeUnit.MINUTES);
if(!completed){
   logCallback.close(); //can not close the channel since logCallback.stream hasn't been set
}
```

I moved resultCallback.onStart from channelRead0 to channelRegistered to handle the situation.

---
This change is [Reviewable](https://reviewable.io/reviews/docker-java/docker-java/898)

### CODE REVIEW

**Comment by fengxx** [comment]
CI failed for
```
14:50:27.861 INFO  c.g.d.c.AbstractDockerClientTest - Connecting to Docker server
14:50:29.224 ERROR c.g.d.c.async.ResultCallbackTemplate - Error during callback
com.github.dockerjava.api.exception.NotFoundException: No such image: busybox
14:50:29.229 INFO  c.g.d.c.AbstractDockerClientTest - Pulling image 'busybox'
14:50:30.475 INFO  c.g.d.c.AbstractDockerClientTest - ======================= END OF BEFORETEST =======================
```
any ideas?

File: src/test/java/com/github/dockerjava/core/command/ListContainersCmdImplTest.java [hunk 1]
```diff
@@ -104,7 +104,7 @@ public void testListContainersWithLabelsFilter() throws Exception {
         String testImage = "busybox";
 
         // need to block until image is pulled completely
-        dockerClient.pullImageCmd(testImage).exec(new PullImageResultCallback()).awaitCompletion();
+        dockerClient.pullImageCmd(testImage).withTag("latest").exec(new PullImageResultCallback()).awaitCompletion();
```
Comment (KostyaSha) [inline src/test/java/com/github/dockerjava/core/command/ListContainersCmdImplTest.java]: without it it downloads all tags?

Comment (fengxx) [inline src/test/java/com/github/dockerjava/core/command/ListContainersCmdImplTest.java]: yes, more obvious on a slow network, and it is documented https://docs.docker.com/engine/api/v1.24/
```
tag – Tag or digest. If empty when pulling an image, this causes all tags for the given image to be pulled.```

Comment (KostyaSha) [inline src/test/java/com/github/dockerjava/core/command/ListContainersCmdImplTest.java]: seems we need review other tests...

**Comment by fengxx** [comment]
use ReadTimeoutHandler caused side effects on HttpResponseStreamHandler which relies on channelInactive,  I will modify the PR to address the easy part: make sure channel can be closed correctly by calling resultCallback.close

**Comment by KostyaSha** [comment]
> will modify the PR to address the easy part: make sure channel can be closed correctly by calling resultCallback.close

Tests are very welcome!

**Comment by KostyaSha** [comment]
push non existing image test is infinitely hanging for me, even when docker daemon answered and closed connection .awaitCompletion() still waits. Is it related?

**Comment by fengxx** [comment]
I am not sure, I tested on localhost and it passed
```
13:13:54.477 DEBUG c.g.d.n.NettyDockerCmdExecFactory$1 - [id: 0x78612ba6, L:/127.0.0.1:50102 - R:localhost/127.0.0.1:2375] WRITE: DefaultHttpRequest(decodeResult: success, version: HTTP/1.1)
POST /images/user/xxx/push HTTP/1.1
content-length: 0
host: 
connection: keep-alive
accept-encoding: gzip
X-Registry-Auth: bnVsbA==
accept: application/json
13:13:54.481 DEBUG c.g.d.n.NettyDockerCmdExecFactory$1 - [id: 0x78612ba6, L:/127.0.0.1:50102 - R:localhost/127.0.0.1:2375] FLUSH
13:13:54.485 DEBUG c.g.d.n.NettyDockerCmdExecFactory$1 - [id: 0x78612ba6, L:/127.0.0.1:50102 - R:localhost/127.0.0.1:2375] READ: DefaultHttpResponse(decodeResult: success, version: HTTP/1.1)
HTTP/1.1 200 OK
Api-Version: 1.30
Content-Type: application/json
Date: Thu, 28 Sep 2017 05:13:54 GMT
Docker-Experimental: true
Ostype: linux
Server: Docker/17.06.0-ce (linux)
Transfer-Encoding: chunked
13:13:54.485 DEBUG c.g.d.n.NettyDockerCmdExecFactory$1 - [id: 0x78612ba6, L:/127.0.0.1:50102 - R:localhost/127.0.0.1:2375] READ: DefaultHttpContent(data: PooledSlicedByteBuf(ridx: 0, widx: 227, cap: 227/227, unwrapped: PooledUnsafeDirectByteBuf(ridx: 445, widx: 445, cap: 1024)), decoderResult: success), 227B
//removed the hexdump
13:13:54.517 DEBUG c.g.d.c.c.PushImageResultCallback - PushResponseItem[stream=<null>,status=The push refers to a repository [docker.io/user/xxx],progressDetail=<null>,progress=<null>,id=<null>,from=<null>,time=<null>,errorDetail=<null>,error=<null>,aux=<null>]
13:13:54.518 DEBUG c.g.d.c.c.PushImageResultCallback - PushResponseItem[stream=<null>,status=<null>,progressDetail=<null>,progress=<null>,id=<null>,from=<null>,time=<null>,errorDetail=ResponseItem.ErrorDetail[code=<null>,message=An image does not exist locally with the tag: user/xxx],error=An image does not exist locally with the tag: user/xxx,aux=<null>]
13:13:54.518 DEBUG c.g.d.n.NettyDockerCmdExecFactory$1 - [id: 0x78612ba6, L:/127.0.0.1:50102 - R:localhost/127.0.0.1:2375] READ COMPLETE
13:13:54.519 DEBUG c.g.d.n.NettyDockerCmdExecFactory$1 - [id: 0x78612ba6, L:/127.0.0.1:50102 - R:localhost/127.0.0.1:2375] READ, EmptyLastHttpContent, 0B
13:13:54.519 DEBUG c.g.d.n.NettyDockerCmdExecFactory$1 - [id: 0x78612ba6, L:/127.0.0.1:50102 - R:localhost/127.0.0.1:2375] CLOSE
13:13:54.520 DEBUG c.g.d.n.NettyDockerCmdExecFactory$1 - [id: 0x78612ba6, L:/127.0.0.1:50102 ! R:localhost/127.0.0.1:2375] READ COMPLETE
13:13:54.520 DEBUG c.g.d.n.NettyDockerCmdExecFactory$1 - [id: 0x78612ba6, L:/127.0.0.1:50102 ! R:localhost/127.0.0.1:2375] INACTIVE
13:13:54.520 DEBUG c.g.d.n.NettyDockerCmdExecFactory$1 - [id: 0x78612ba6, L:/127.0.0.1:50102 ! R:localhost/127.0.0.1:2375] UNREGISTERED
```

**Comment by KostyaSha** [comment]
I see different behaviour on 1.26 API

**Comment by KostyaSha** [comment]
Dropping out 1.10 support :D

**Comment by KostyaSha** [comment]
sorry, is it done?

**Comment by fengxx** [comment]
the PR is outdated, just closing it

**Comment by KostyaSha** [comment]
It has no conflicts, is it still needed?

**Pull request closed without merging**